code/game/objects/items/weapons/thurible.dm

```diff
@@ -0,0 +1,169 @@
+		put_out(user)
+	return ..()
+
+/obj/item/thurible/can_enter_storage(obj/item/storage/S, mob/user)
+	if(lit)
+		to_chat(user, "<span class='warning'>[S] can't hold \the [initial(name)] while it's lit!</span>") // initial(name) so it doesn't say "lit" twice in a row
+		return FALSE
+	return TRUE
+
+/obj/item/thurible/fire_act(datum/gas_mixture/air, exposed_temperature, exposed_volume, global_overlay = TRUE)
+	. = ..()
+	light()
+
+/obj/item/thurible/on_reagent_change()
+	. = ..()
+	if(!corrupted)
+		var/found_forbidden_reagent = FALSE
+		for(var/datum/reagent/R in reagents.reagent_list)
+			if(R.id == "unholywater")
+				corrupted = TRUE
+				to_chat(loc, "<span class='warning'>[src] is corrupted by an unholy substance!</span>")
+				return
+			if(!safe_chem_list.Find(R.id))
+				reagents.del_reagent(R.id)
+				found_forbidden_reagent = TRUE
+		if(found_forbidden_reagent)
+			if(ismob(loc))
+				to_chat(loc, "<span class='warning'>[src] banishes a dangerous substance!</span>")
+			else
+				visible_message("<span class='warning'>[src] banishes a dangerous substance!</span>")
+
+/// Lights the thurible and starts processing reagents
+/obj/item/thurible/proc/light(mob/user)
+	if(lit)
+		to_chat(user, "<span class='warning'>The [src] is already lit!</span>")
+		return
+
+	if(!reagents.total_volume)
+		to_chat(user, "<span class='warning'>The [src] is out of fuel!</span>")
+		return
+
+	// Plasma explodes when exposed to fire.
+	if(reagents.get_reagent_amount("plasma"))
+		var/datum/effect_system/reagents_explosion/E = new()
+		E.set_up(round(reagents.get_reagent_amount("plasma") / 2.5, 1), get_turf(src), 0, 0)
+		E.start()
+		return
+
+	// Fuel explodes, too, but much less violently.
+	if(reagents.get_reagent_amount("fuel"))
+		var/datum/effect_system/reagents_explosion/E = new()
+		E.set_up(round(reagents.get_reagent_amount("fuel") / 5, 1), get_turf(src), 0, 0)
+		E.start()
+		return
+
+	// And black powder... but more violently.
+	if(reagents.get_reagent_amount("blackpowder"))
+		var/datum/effect_system/reagents_explosion/E = new()
+		E.set_up(round(reagents.get_reagent_amount("blackpowder") / 2, 1), get_turf(src), 0, 0)
+		E.start()
+		return
+
+	lit = TRUE
+	reagents.set_reacting(TRUE)
+	reagents.handle_reactions()
+	START_PROCESSING(SSobj, src)
+	set_light(2, 0.3, "#E38F46")
+	update_appearance()
+	return TRUE
+
+/// Extinguishes the thurible and stops processing
+/obj/item/thurible/proc/put_out(mob/user)
+	lit = FALSE
+	STOP_PROCESSING(SSobj, src)
+	set_light(0)
+	update_appearance()
+	return TRUE
+
+/// Spreads reagents in a 3x3 area centered on the thurible
+/obj/item/thurible/proc/swing()
+	var/obj/released_reagents = new
+	released_reagents.create_reagents(2)
+	reagents.trans_to(released_reagents, swing_reagents_consumed)
+	var/list/mobs_to_smoke = list()
+	var/list/smoked_atoms = list()
+	for(var/atom/A in view(1, get_turf(src.loc)))
+		if(A in smoked_atoms)
+			continue
+		smoked_atoms += A
```

Comment:
Do you use `smoked_atoms` for anything here beyond checking for duplicates? I'm not entirely sure if you really need that

Reply:
It's not used for anything aside from ensuring the same atom is not exposed to the reagent twice.

Reply:
Testing with that removed found that mobs were receiving the correct amount of reagents from the device. I have removed it as it is extraneous.